components/hab/src/cli.rs

```diff
@@ -202,6 +202,38 @@ pub fn get(feature_flags: FeatureFlag) -> App<'static, 'static> {
                 (about: "Commands relating to Habitat Builder channels")
                 (aliases: &["c", "ch", "cha", "chan", "chann", "channe"])
                 (@setting ArgRequiredElseHelp)
+                (@subcommand promote =>
+                    (about: "Atomically promotes all packages in channel")
+                    (@arg BLDR_URL: -u --url +takes_value {valid_url}
+                        "Specify an alternate Builder endpoint. If not specified, the value will \
+                         be taken from the HAB_BLDR_URL environment variable if defined. (default: \
+                         https://bldr.habitat.sh)")
+                    (@arg ORIGIN: -o --origin +required +takes_value {valid_origin}
+                        "The origin for the channels. Default is from \
+                        'HAB_ORIGIN' or cli.toml")
+                    (@arg SOURCE_CHANNEL: +required +takes_value
+                        "The channel from which all packages will be selected for promotion")
+                    (@arg TARGET_CHANNEL: +required +takes_value
+                        "The channel to which packages will be promoted")
+                    (@arg AUTH_TOKEN: -z --auth +takes_value "Authentication token for Builder")
+
+                )
+                (@subcommand demote =>
+                    (about: "Atomically demotes selected packages in a target channel")
```

Comment:
I tried testing this out and got the following error:
```
➤ cargo run -p hab bldr channel demote --origin baumanj bldr-1221143477681143808 bldr-→ Demoting Packages selected from channel bldr-1221143477681143808 that are residing in bldr-1221143477681143808.
[2019-08-30T01:52:39Z DEBUG reqwest::async_impl::response] Response: '404 Not Found' for https://bldr.habitat.sh/v1/depot/channels/baumanj/bldr-1221143477681143808/pkgs/demote?channel=bldr-1221143477681143808
✗✗✗
✗✗✗ [404 Not Found]
✗✗✗
```
If that's operating like it should, I think we need a more informative error for the user. However, I think maybe it's not operating like it should, because when I try demoting from a different channel, I get the same error:
```
➤ cargo run -p hab bldr channel demote --origin baumanj bldr-1221143477681143808 stable
→ Demoting Packages selected from channel bldr-1221143477681143808 that are residing in stable.
[2019-08-30T01:53:24Z DEBUG reqwest::async_impl::response] Response: '404 Not Found' for https://bldr.habitat.sh/v1/depot/channels/baumanj/bldr-1221143477681143808/pkgs/demote?channel=stable
✗✗✗
✗✗✗ [404 Not Found]
✗✗✗
```

Am I invoking this wrongly? I also got a 404 with `hab bldr channel promote --origin baumanj bldr-1221143477681143808 foo`.

Reply:
This API isn't in live yet, so yes that is what I would expect. I'm not even certain it's in acceptance yet.

Reply:
I'm coming into this late, so forgive me if this has been covered already, but why did we merge these CLI changes if the API to support them doesn't exist?  Wouldn't it make more sense to get the server API created first and verify that it works?  Are our users going to be running commands that don't do anything?

Reply:
@raskchanky the API exists and is on acceptance - it will be propped to live on Tue

Reply:
Ah, thank you for clarifying.

Reply:
I'm getting behavior I don't understand with acceptance. I tried promoting to create a new channel `foo`:
```sh
➤ cargo run -p hab bldr channel promote -z "$HAB_AUTH_TOKEN_ACCEPTANCE" --url "https://bldr.acceptance.habitat.sh" --origin baumanj bldr-1027558301272449024 foo
→ Promoting Packages from channel bldr-1027558301272449024 to foo.
[2019-08-30T19:53:54Z DEBUG reqwest::async_impl::response] Response: '200 OK' for https://bldr.acceptance.habitat.sh/v1/depot/channels/baumanj/bldr-1027558301272449024/pkgs/promote?channel=foo
✓ Promoted packages from channel bldr-1027558301272449024 to foo.
```
That appeared to work, though I'm currently seeing inconsistent results when I refresh https://bldr.acceptance.habitat.sh/#/pkgs/baumanj/minimial-habitat-app/0.1.0/20180716163529 versus https://bldr.acceptance.habitat.sh/#/pkgs/baumanj/minimial-habitat-app whether baumanj/minimial-habitat-app/0.1.0/20180716170506 (which is the one that is in the `bldr-1027558301272449024` channel) is also a member of `bar`:

![Screen Shot 2019-08-30 at 1 05 13 PM](https://user-images.githubusercontent.com/5906042/64048540-db4d9000-cb26-11e9-920b-40d5b48dd5f0.png)
![Screen Shot 2019-08-30 at 1 04 48 PM](https://user-images.githubusercontent.com/5906042/64048541-db4d9000-cb26-11e9-93a5-a7c1ae989ff6.png)

Reply:
Additionally, demoting with the same source and target channel fails:
```sh
➤ cargo run -p hab bldr channel demote -z "$HAB_AUTH_TOKEN_ACCEPTANCE" --url "https://bldr.acceptance.habitat.sh" --origin baumanj bar bar
→ Demoting Packages selected from channel bar that are residing in bar.
[2019-08-30T20:06:03Z DEBUG reqwest::async_impl::response] Response: '400 Bad Request' for https://bldr.acceptance.habitat.sh/v1/depot/channels/baumanj/bar/pkgs/demote?channel=bar
✗✗✗
✗✗✗ [400 Bad Request]
✗✗✗
```
but I think it would be better to give a more informative message to the user

Reply:
What I'm seeing in the audit logs is a promotion to channel foo, a demotion from channel foo, and then a promotion to channel bar. Which would be consistent with the current state of that package. As far as the message output, we can definitely iterate on the error handling to get it cleared up.

Reply:
I'm thinking the above inconsistent results may be a stale data in UX issue, but I'm really not sure. @apriofrost or @chefsalim, have you ever seen anything like that?

Reply:
@baumanj I'm not sure what you mean by "inconsistent results"? https://bldr.acceptance.habitat.sh/#/pkgs/baumanj/minimial-habitat-app/0.1.0/20180716163529 this pkg is a non-latest package that doesn't belong to `bldr-1027558301272449024` channel, so it's not in the `bar` channel.  Though there is an inconsistency in how the channel badges are sorted sequentially, seems to depend on which previous page you land before the `versions` tab, that's a known UI issue:
**sorting order 1**
![image](https://user-images.githubusercontent.com/9401459/64197513-20313980-ce3b-11e9-8f30-d2a3b18ec8cc.png)
**sorting order 2**
![image](https://user-images.githubusercontent.com/9401459/64197624-5b336d00-ce3b-11e9-97d3-880144959fc6.png)

Reply:
I think this was my mistake. Sorry about that, @apriofrost.

Reply:
no worries! thanks for paying attention to the details @baumanj !